cpp/src/traversal/sssp.cu

```diff
@@ -0,0 +1,430 @@
+
+#include "traversal_common.cuh"
+#include "sssp.cuh"
+#include "sssp_kernels.cuh"
+#include "utilities/error_utils.h"
+
+namespace cugraph {
+
+template <typename IndexType, typename DistType>
+void SSSP<IndexType, DistType>::setup() {
+  // Working data
+  // Each vertex can be in the frontier at most once
+  ALLOC_TRY(&frontier, n * sizeof(IndexType), nullptr);
+  ALLOC_TRY(&new_frontier, n * sizeof(IndexType), nullptr);
+
+  // size of bitmaps for vertices
+  vertices_bmap_size = (n / (8 * sizeof(int)) + 1);
+
+  // size of bitmaps for edges
+  edges_bmap_size = (nnz / (8 * sizeof(int)) + 1);
+
+  // ith bit of isolated_bmap is set <=> degree of ith vertex = 0
+  ALLOC_TRY(&isolated_bmap, sizeof(int) * vertices_bmap_size, nullptr);
+
+  // Allocate buffer for data that need to be reset every iteration
+  iter_buffer_size =
+      sizeof(int) * (edges_bmap_size + vertices_bmap_size) + sizeof(IndexType);
+  ALLOC_TRY(&iter_buffer, iter_buffer_size, nullptr);
+  // ith bit of relaxed_edges_bmap <=> ith edge was relaxed
+  relaxed_edges_bmap = (int*)iter_buffer;
+  // ith bit of next_frontier_bmap <=> vertex is active in the next frontier
+  next_frontier_bmap = (int*)iter_buffer + edges_bmap_size;
+  // num vertices in the next frontier
+  d_new_frontier_cnt = next_frontier_bmap + vertices_bmap_size;
+
+  // vertices_degree[i] = degree of vertex i
+  ALLOC_TRY(&vertex_degree, sizeof(IndexType) * n, nullptr);
+
+  // Cub working data
+  traversal::cub_exclusive_sum_alloc(
+      n + 1, d_cub_exclusive_sum_storage, cub_exclusive_sum_storage_bytes);
+
+  // frontier_vertex_degree[i] is the degree of vertex frontier[i]
+  ALLOC_TRY(&frontier_vertex_degree, n * sizeof(IndexType), nullptr);
+
+  // exclusive sum of frontier_vertex_degree
+  ALLOC_TRY(&exclusive_sum_frontier_vertex_degree,
+            (n + 1) * sizeof(IndexType),
+            nullptr);
+
+  // We use buckets of edges (32 edges per bucket for now, see exact macro in
+  // sssp_kernels). frontier_vertex_degree_buckets_offsets[i] is the index k
+  // such as frontier[k] is the source of the first edge of the bucket
+  // See top down kernels for more details
+  size_t bucket_off_size =
+      ((nnz / TOP_DOWN_EXPAND_DIMX + 1) * NBUCKETS_PER_BLOCK + 2) *
+      sizeof(IndexType);
+  ALLOC_TRY(&exclusive_sum_frontier_vertex_buckets_offsets,
+            bucket_off_size,
+            nullptr);
+
+  // Repurpose d_new_frontier_cnt temporarily
+  IndexType* d_nisolated = d_new_frontier_cnt;
+  cudaMemsetAsync(d_nisolated, 0, sizeof(IndexType), stream);
+
+  // Computing isolated_bmap
+  // Only dependent on graph - not source vertex - done once
+  traversal::flag_isolated_vertices(
+      n, isolated_bmap, row_offsets, vertex_degree, d_nisolated, stream);
+
+  cudaMemcpyAsync(&nisolated,
+                  d_nisolated,
+                  sizeof(IndexType),
+                  cudaMemcpyDeviceToHost,
+                  stream);
+
+  // We need nisolated to be ready to use
```

Comment:
What is nisolated???

Reply:
nisolated is the number of isolated vertices, where isolated is defined as vertices with zero out-degree. The naming is a bit misleading, but I kept it to keep it consistent with the bfs codebase. I can also change it uniformly to something else across bfs and sssp.

Reply:
Ah... OK, I mistakenly thought it's a typo.